### PR TITLE
fix rewind crash on MSVC build when using SSE2

### PR DESCRIPTION
**Pull request by Jamiras**
## Description

Fixes a crash using rewind on Windows with SSE2 enabled.

`state_manager_raw_compress` writes out three 16-bit values when it detects no changes in more than 64KB https://github.com/libretro/RetroArch/blob/2520337ddcc9923acbd50473f0a7ea473f8f1e2b/state_manager.c#L258-L260

but only two 16-bit values when it does find a change https://github.com/libretro/RetroArch/blob/2520337ddcc9923acbd50473f0a7ea473f8f1e2b/state_manager.c#L268-L269

The problem is that the `find_change` logic was not returning the correct offset, so it would sometimes write a `0` for the number of changed items in a sequence, and then one 16-bit value for the number of non-changed items. 

Having a `0` for the number of changed items tells `state_manager_raw_decompress` to read two 16-bit values to determine the number of non-changed items. Since only one was output, when two values are read, the count is often much larger than the size of the save state buffer and the next pair of values causes memory access violations and the application crashes.

This is caused by `compat_ctz` (count trailing zeroes) calling the wrong function. `_BitScanReverse` returns the number of leading zeros, not the number of trailing zeroes. 

https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanreverse-bitscanreverse64?view=msvc-160
> Search the mask data from most significant bit (MSB) to least significant bit (LSB) for a set bit (1).

Trailing zeroes are the least significant bits, so the code should be calling `_BitScanForward`.

I've also made a minor change in `find_change` itself. By calling `_mm_cmpeq_epi8` instead of `_mm_cmpeq_epi32`, we can leverage the SIMD instructions for finding the exact byte that changed instead of just identifying which 32-bit value changed and doing an additional check to see if it was the high or low 16-bit value that changed. For completeness, I've updated the fallback implementation of `compat_ctz` to support full granularity of all 32-bits, whereas it previously only supported 0, 4, 8, 12, and 16.

## Related Issues

n/a

## Related Pull Requests

n/a

## Reviewers

@twinaphex

### CODE REVIEW

File: libretro-common/include/compat/intrinsics.h [hunk 1]
```diff
@@ -63,20 +63,34 @@ static INLINE int compat_ctz(unsigned x)
    return __builtin_ctz(x);
 #elif _MSC_VER >= 1400 && !defined(_XBOX) && !defined(__WINRT__)
    unsigned long r = 0;
-   _BitScanReverse((unsigned long*)&r, x);
+   _BitScanForward((unsigned long*)&r, x);
```
Comment (Jamiras) [inline libretro-common/include/compat/intrinsics.h:66]: This is the bugfix. I believe the other changes are improvements, but they are not required if you'd prefer I change them back.

**Comment by twinaphex** [comment]
Thanks, I will ask @Alcaro if he wants to review this one since he wrote a large part of the rewind code.

**Comment by Alcaro** [comment]
Neat, that is one ancient bug. I never tested on MSVC, I just used the first google hit and misread something. Maybe that's why it's excluded on xbox. (No clue why the GCC version is disabled on console, though...)

The epi32 is because I didn't know SSE when I wrote that - or rather, I only knew a few instructions and didn't know where to find docs on the rest. Fixing that sounds good to me.

The old scalar compat_ctz is enough for this function, but such a limited thing doesn't belong in -common.

In short, all of this looks good to me.

**Comment by twinaphex** [comment]
Cool to hear this resolves some longstanding issues.
